lib/postgrex/protocol.ex

```diff
@@ -99,39 +101,76 @@ defmodule Postgrex.Protocol do
       types_lock: nil,
       prepare: prepare,
       messages: [],
-      ssl: ssl?
+      ssl: ssl?,
+      target_server_type: target_server_type,
+      remaining_endpoints: remaining_endpoints_to_attempt
     }
 
-    case connect(host, port, sock_opts ++ @sock_opts, connect_timeout, s) do
-      {:ok, s} -> handshake(s, status)
-      {:error, _} = error -> error
-    end
+    connect_endpoints(host, port, sock_opts ++ @sock_opts, connect_timeout, s, status)
   end
 
-  defp host_and_port(opts) do
+  defp host_and_port_params(opts) do
     port = opts[:port] || 5432
 
     case Keyword.fetch(opts, :socket) do
       {:ok, file} ->
-        {{:local, file}, 0}
+        {{:local, file}, 0, nil}
 
       :error ->
         case Keyword.fetch(opts, :socket_dir) do
           {:ok, dir} ->
-            {{:local, "#{dir}/.s.PGSQL.#{port}"}, 0}
+            {{:local, "#{dir}/.s.PGSQL.#{port}"}, 0, nil}
```

Comment:
Make it return an empty list when there is no end-point, this will simplify the code later on. :)